src/Symfony/Component/Validator/Constraints/FileValidator.php

```diff
@@ -168,19 +168,61 @@ public function validate(mixed $value, Constraint $constraint)
             }
         }
 
-        if ($constraint->mimeTypes) {
+        $mimeTypes = (array) $constraint->mimeTypes;
+
+        if ($constraint->extensions) {
+            $fileExtension = pathinfo($path, \PATHINFO_EXTENSION);
+
+            $found = false;
+            $normalizedExtensions = [];
+            foreach ((array) $constraint->extensions as $k => $v) {
+                if (!\is_string($k)) {
+                    $k = $v;
+                    $v = null;
+                }
+
+                $normalizedExtensions[] = $k;
+
+                if ($fileExtension !== $k) {
+                    continue;
+                }
+
+                $found = true;
+                if (null === $v) {
+                    if (!class_exists(MimeTypes::class)) {
+                        throw new LogicException('You cannot validate the mime-type of files as the Mime component is not installed. Try running "composer require symfony/mime".');
+                    }
+
+                    $mimeTypesHelper = MimeTypes::getDefault();
+                    $v = $mimeTypesHelper->getMimeTypes($k);
+                }
+
+                $mimeTypes = $mimeTypes ? array_intersect($v, $mimeTypes) : (array) $v;
```

Comment:
the array cast is not needed, isn't it?

Reply:
It is, this allows this syntax `['xml' => 'text/xml']`.

Reply:
PHPDoc fixed to document that.